File: DataStructureAndAlgorithm/Greedy/HuffmanCoding/C++/HuffmanCoding.cpp
#include "HuffmanCoding.h"
#include <iostream>

void printCodes(MinHeapNode* root, std::string str) {
    if (!root) return;

    if (root->data != '$') {
        std::cout << root->data << ": " << str << std::endl;
    }

    printCodes(root->left, str + "0");
    printCodes(root->right, str + "1");
}

void HuffmanCodes(const std::string& data, const std::vector<int>& freq) {
    MinHeapNode *left, *right, *top;

    std::priority_queue<MinHeapNode*, std::vector<MinHeapNode*>, compare> minHeap;

    for (size_t i = 0; i < data.length(); ++i) {
        minHeap.push(new MinHeapNode(data[i], freq[i]));
    }

    while (minHeap.size() != 1) {
        left = minHeap.top();
        minHeap.pop();

        right = minHeap.top();
        minHeap.pop();

        top = new MinHeapNode('$', left->freq + right->freq);
        top->left = left;
        top->right = right;
        minHeap.push(top);
    }

    printCodes(minHeap.top(), "");
}

File: DataStructureAndAlgorithm/Greedy/HuffmanCoding/Python/huffman_coding.py
import heapq

class Node:
    def __init__(self, char, freq):
        self.char = char
        self.freq = freq
        self.left = None
        self.right = None

    # Define comparison for priority queue
    def __lt__(self, other):
        return self.freq < other.freq

def huffman_coding(data):
    # data is a dictionary of character frequencies: {'a': 5, 'b': 9, ...}

    priority_queue = []
    for char, freq in data.items():
        heapq.heappush(priority_queue, Node(char, freq))

    while len(priority_queue) > 1:
        left = heapq.heappop(priority_queue)
        right = heapq.heappop(priority_queue)

        merged = Node(None, left.freq + right.freq)
        merged.left = left
        merged.right = right

        heapq.heappush(priority_queue, merged)

    # The single node left in the priority queue is the root of the Huffman tree
    root = heapq.heappop(priority_queue)

    codes = {}
    def build_codes(node, current_code):
        if node is None: return

        if node.char is not None:
            codes[node.char] = current_code
            return

        build_codes(node.left, current_code + "0")
        build_codes(node.right, current_code + "1")

    build_codes(root, "")
    return codes
